wildcards_test.R

```diff
@@ -0,0 +1,281 @@
+  ),
+  display_dropdown <- function(n_clicks, children){
+    new_element = htmlDiv(list(
+      dccDropdown(
+        id = list("index" = n_clicks, "type" = "dynamic-dropdown"),
+        options = lapply(c("NYC", "MTL", "LA", "TOKYO"), function(x){
+          list("label" = x, "value" = x)
+        })
+      ),
+      htmlDiv(
+        id = list("index" = n_clicks, "type" = "dynamic-output"),
+        children = list()
+      )
+    ))
+    
+    children <- c(children, list(new_element))
+    return(children)
+  }
+)
+
+
+app$callback(
+  output(id = list("index" = dash:::MATCH, "type" = "dynamic-output"), property= "children"),
+  params = list(
+    input(id=list("index" = dash:::MATCH, "type" = "dynamic-dropdown"), property= "value"),
+    state(id=list("index" = dash:::MATCH, "type" = "dynamic-dropdown"), property= "id")
+  ),
+  display_output <- function(value, id){
+    # ctx <- app$callback_context()
+    # print(jsonlite::toJSON(ctx, pretty = TRUE))
+    print(id)
+    print(value)
+    return(htmlDiv(sprintf("Dropdown %s = %s", id$index, value)))
+  }
+)
+
+app$run_server(showcase = T, debug = T, port = 8070)
+  
+
+
+# Example with ALLSMALLER 
+
+
+library(dash)
+library(dashCoreComponents)
+library(dashHtmlComponents)
+
+
+df <- read.csv('https://raw.githubusercontent.com/plotly/datasets/master/gapminder2007.csv', stringsAsFactors = F)
+
+
+app <- Dash$new()
+
+app$layout(htmlDiv(list(
+  htmlButton("Add Filter", id = "add-filter-ex3", n_clicks = 0),
+  htmlDiv(id = "container-ex3", children = list())
+)))
+
+
+app$callback(
+  output('container-ex3', 'children'),
+  params = list(
+    input('add-filter-ex3', 'n_clicks'),
+    state('container-ex3', 'children')
+  ),
+  display_dropdowns <- function(n_clicks, existing_children){
+    new_children <- htmlDiv(list(
+      dccDropdown(
+        id = list("index" = n_clicks, "type" = "filter-dropdown-ex3"),
+        options = lapply(unique(df$country), function(x){
+          list("label" = x, "value" = x)
+        }),
+        value = unique(df$country)[n_clicks + 1]
+      ),
+      htmlDiv(id = list("index" = n_clicks, "type" = "output-ex3"), children = list(unique(df$country)[n_clicks + 1]))
+    ))
+    
+    existing_children <- c(existing_children, list(new_children))
+  }
+)
+
+
+app$callback(
+  output(id = list("type" = "output-ex3", "index" = dash:::MATCH), property = "children"),
+  params = list(
+    input(id = list("type" = "filter-dropdown-ex3", "index" = dash:::MATCH), property = "value"),
+    input(id = list("type" = "filter-dropdown-ex3", "index" = dash:::ALLSMALLER), property = "value")
+  ),
+  display_output <- function(matching_value, previous_values){
+    previous_values_in_reversed_order = rev(previous_values)
+    all_values = c(matching_value, previous_values_in_reversed_order)
+    all_values = unlist(all_values)
+    
+    dff = df[df$country %in% all_values,]
+    avgLifeExp = round(mean(dff$lifeExp), digits = 2)
```

Comment:
Not needed here (no missing values), but (something to think about in the future) I've gotten in the habit of using `mean(x, na.rm=TRUE)` since there are often missing values and I'd rather have an estimate than an `NA`. :laughing: